src/wasm-type.h

```diff
@@ -44,36 +51,45 @@ class Type {
     externref,
     nullref,
     exnref,
-    _last_value_type = exnref
+    _last_basic_id = exnref
   };
 
   Type() = default;
 
-  // ValueType can be implicitly upgraded to Type
-  constexpr Type(ValueType id) : id(id){};
+  // BasicID can be implicitly upgraded to Type
+  constexpr Type(BasicID id) : id(id){};
 
   // But converting raw uint32_t is more dangerous, so make it explicit
   explicit Type(uint64_t id) : id(id){};
 
-  // Construct from lists of elementary types
-  Type(std::initializer_list<Type> types);
-  explicit Type(const std::vector<Type>& types);
+  // Construct tuple from lists of elementary types
+  Type(std::initializer_list<Type>);
+
+  // Construct from tuple description
+  explicit Type(const Tuple&);
+
+  // Construct from signature description
+  explicit Type(const Signature, bool nullable);
+
+  // Construct from struct description
+  explicit Type(const Struct&, bool nullable);
+
+  // Construct from array description
+  explicit Type(const Array&, bool nullable);
 
   // Accessors
   size_t size() const;
-  const std::vector<Type>& expand() const;
+  const TypeList& expand() const;
 
   // Predicates
-  constexpr bool isSingle() const {
-    return id >= i32 && id <= _last_value_type;
-  }
-  constexpr bool isMulti() const { return id > _last_value_type; }
+  constexpr bool isSingle() const { return id >= i32 && id <= _last_basic_id; }
   constexpr bool isConcrete() const { return id >= i32; }
   constexpr bool isInteger() const { return id == i32 || id == i64; }
   constexpr bool isFloat() const { return id == f32 || id == f64; }
   constexpr bool isVector() const { return id == v128; };
   constexpr bool isNumber() const { return id >= i32 && id <= v128; }
-  constexpr bool isRef() const { return id >= funcref && id <= exnref; }
+  bool isTuple() const;
+  bool isRef() const;
```

Comment:
I think these can be `constexpr` as well.

Reply:
That's one more of these cases where I don't know how to resolve another problem. If I make these `constexpr`, I need to place their implementation in the header file, and since they make use of `TypeDef` I then have to also forward-declare `TypeDef`, incl. forward-declaring its `constexpr`s used in `isRef` etc., or I get errors complaining about an incomplete type. Everything I tried just resulted in more headaces, so I gave up :)

Reply:
Hmm yeah gross. Definitely not worth jumping through hoops for.